### PR TITLE
Added Protection Policy CRD operations

### PR DESCRIPTION
**Pull request by prajwalpatil25**
# Description
Added Create, Read and Delete functionalities for protection policy. Also, added example for protection policy in main.tf

# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
| |

# Checklist:

- [X] I have performed a self-review of my own code to ensure there are no formatting, vetting, linting, or security issues
- [ ] I have verified that new and existing unit tests pass locally with my changes
- [ ] I have not allowed coverage numbers to degenerate
- [ ] I have maintained at least 90% code coverage
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [X] Backward compatibility is not broken

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Please also list any relevant details for your test configuration
Tested execution of main.tf. Protection Policy was created and destroyed successfully.
[protection policy CRD testing.txt](https://github.com/dell/terraform-provider-powerstore/files/10224241/protection.policy.CRD.testing.txt)

### CODE REVIEW

File: models/protectionpolicy.go [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+	ID                 types.String `tfsdk:"id"`
+	Name               types.String `tfsdk:"name"`
+	Description        types.String `tfsdk:"description"`
+	Type               types.String `tfsdk:"type"`
+	ManagedBy          types.String `tfsdk:"managed_by"`
+	ManagedByID        types.String `tfsdk:"managed_by_id"`
+	IsReadOnly         types.Bool   `tfsdk:"is_read_only"`
+	IsReplica          types.Bool   `tfsdk:"is_replica"`
+	SnapshotRuleIds    types.List   `tfsdk:"snapshot_rule_ids"`
+	ReplicationRuleIds types.List   `tfsdk:"replication_rule_ids"`
```
Comment (mkmayank-dell) [inline models/protectionpolicy.go]: rename SnapshotRuleIds    as SnapshotRuleIDs
same for ReplicationRuleIds as ReplicationRuleIDs

File: powerstore/resource_protection_policy.go [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Error reading protection policy",
+			"Could not read protection policy with error "+id+": "+err.Error(),
+		)
+		return
+	}
+
+	updatePolicyState(&state, response, &state)
```
Comment (mkmayank-dell) [inline powerstore/resource_protection_policy.go]: code is missing for setting the state back to terraform

File: powerstore/resource_protection_policy.go [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+}
+
+// Update updates protection policy
+func (r resourceProtectionPolicy) Update(ctx context.Context, req tfsdk.UpdateResourceRequest, resp *tfsdk.UpdateResourceResponse) {
+}
+
+func planToProtectionPolicyParam(plan models.ProtectionPolicy) *gopowerstore.ProtectionPolicyCreate {
```
Comment (mkmayank-dell) [inline powerstore/resource_protection_policy.go]: Please move this function to resourceProtectionPolicy's method

File: powerstore/resource_protection_policy.go [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+	protectionPolicyCreate := &gopowerstore.ProtectionPolicyCreate{
+		Name:               plan.Name.Value,
+		Description:        plan.Description.Value,
+		ReplicationRuleIds: replicationRuleIds,
+		SnapshotRuleIds:    snapshotRuleIds,
+	}
+	return protectionPolicyCreate
+}
+
+func updatePolicyState(polState *models.ProtectionPolicy, polResponse gopowerstore.ProtectionPolicy, polPlan *models.ProtectionPolicy) {
```
Comment (mkmayank-dell) [inline powerstore/resource_protection_policy.go]: Please move this function to resourceProtectionPolicy's method

**Review by mkmayank-dell — APPROVED** [review]

File: examples/main.tf [hunk 5]
```diff
@@ -82,3 +82,11 @@ resource "powerstore_storagecontainer" "test" {
   storage_protocol = "SCSI"
   high_water_mark = 70
 }
+
+resource "powerstore_protectionpolicy" "terraform-provider-test" {
+  # (resource arguments)
+  description = "Creating Protection Policy"
+  name = "test_protection_policy"
+  #replication_rule_ids = ["a50a222f-3ac6-4168-b3d7-6c14b5874c61"]
+  snapshot_rule_ids = ["153df6eb-3433-4b5e-942e-ecf90348df20"]
```
Comment (shenda1) [inline examples/main.tf:91]: Can we implement the feedback we got in the demo i.e supporting names and ids?

File: powerstore/resource_protection_policy.go [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+				Computed:            true,
+				Optional:            true,
+				Description:         "The is_read_only of the protection policy.",
+				MarkdownDescription: "The is_read_only of the protection policy.",
+			},
+			"is_replica": {
+				Type:                types.BoolType,
+				Computed:            true,
+				Optional:            true,
+				Description:         "The is_replica of the protection policy.",
```
Comment (shenda1) [inline powerstore/resource_protection_policy.go]: Please check the description once on the docs

File: powerstore/resource_protection_policy.go [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+				MarkdownDescription: "The is_replica of the protection policy.",
+			},
+			"snapshot_rule_ids": {
+				Type:                types.ListType{ElemType: types.StringType},
+				Computed:            true,
+				Optional:            true,
+				Description:         "The snapshot_rule_ids of the protection policy.",
```
Comment (shenda1) [inline powerstore/resource_protection_policy.go]: The snapshot rules associated with the protection policy?

File: powerstore/resource_protection_policy.go [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+				MarkdownDescription: "The snapshot_rule_ids of the protection policy.",
+			},
+			"replication_rule_ids": {
+				Type:                types.ListType{ElemType: types.StringType},
+				Computed:            true,
+				Optional:            true,
+				Description:         "The replication_rule_ids of the protection policy.",
```
Comment (shenda1) [inline powerstore/resource_protection_policy.go]: The replication rules associated with the protection policy?

File: powerstore/resource_protection_policy.go [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+			"Error deleting protection policy",
+			"Could not delete protectionPolicyID "+protectionPolicyID+": "+err.Error(),
+		)
+		return
+	}
+
+	log.Printf("Done with Delete")
+}
+
+// Read fetch info about asked protection policy
```
Comment (shenda1) [inline powerstore/resource_protection_policy.go]: fetch is required?

File: examples/main.tf (hunk 5 above)
Comment (prajwalpatil25) [reply examples/main.tf:91]: Will implement this in upcoming PR. Can be tracked through https://jira.cec.lab.emc.com/browse/TFM-262

**Review by shenda1 — APPROVED** [review]

**Pull request merged**